### ISSUE TITLE
Secure-Delete.ps1 is not deleting files securely

### ISSUE THREAD

**Issue by Daigle-ISM** [ian8667/powershell#4]
Since a34a32799535e61e73e38077e18550b71cafb6ba Secure-Delete.ps1 is not deleting items securely. Moving an item between volumes is essentially just copying it then deleting the original, which `Move-Item` of course does not do securely.

Below you can see me create a file named test.txt and delete it with the script:

    D:\> get-randomstring -Length 50 |set-content test.txt
    D:\> ls
    
    
        Directory: D:\
    
    
    Mode                 LastWriteTime         Length Name
    ----                 -------------         ------ ----
    -a----        2022-11-25     13:17             52 test.txt
    -a----        2022-11-25     12:32          23501 Secure-Delete.ps1
    
    
    D:\> cat .\test.txt
    IVmNCZ238M9YPegg4YyInw56g9TgGiP8RR3dRgDDAygXXhyxi5
    D:\> .\Secure-Delete.ps1 -Path (Get-Item .\test.txt)
    
    Secure deletion of file
    Today is Friday, 25 November 2022 13:18:13
    Running script Secure-Delete.ps1 in directory D:\
    Remove file
    Are you sure you want to perform this action?
    Performing the operation remove file on target D:\test.txt.
    This action cannot be undone! Please make sure
    [Y] Yes  [N] No  [E] Exit  [?] Help (default is "N"): y
    
    New derived filename: [qr3nb3ua.dfa]
    
    Renaming original file:
    [D:\test.txt]
    to:
    [C:\Users\testuser\AppData\Local\Temp\qr3nb3ua.dfa]
    Length of file to be overwritten: 52 bytes
    
    File overwrite pass #1
    
    File overwrite pass #2
    
    File overwrite pass #3
    Delete-File : Method invocation failed because [System.Security.Cryptography.RandomNumberGenerator] does not contain a
    method named 'GetBytes'.
    At D:\Secure-Delete.ps1:705 char:4
    +    Delete-File -FileName $dest;
    +    ~~~~~~~~~~~~~~~~~~~~~~~~~~~
        + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
        + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Delete-File
    
    D:\>

_C:\Users\testuser\AppData\Local\Temp\qr3nb3ua.dfa still existed at the end, but it had been overwritten, I'm not worried about that right now_

Then I ran Recuva on that drive:
![image](https://user-images.githubusercontent.com/93410187/204056452-9370aad5-e619-427d-9967-41b76df60a5a.png)

And you can see I easily recovered the full contents of that file:
![image](https://user-images.githubusercontent.com/93410187/204056810-83efa48e-b5d6-4447-beb1-30e1261ec679.png)

I would suggest removing the "moving" feature and processing the file in place

Full disclosure: I am not a security expert (nor have I dug any further into the script), so I can't say with confidence that the method being used previously was secure, only that the method being used now defeats the purpose of the script

**Comment by ian8667** [comment]
Hi,
Have you tried using 'Secure-Delete.ps1' version '2022-06-30T11:10:33'? The script was written with PowerShell version 7.3.4 in mind.
Thanks.

**Comment by Daigle-ISM** [comment]
Yes, I'm using the latest version from master [a34a327](https://github.com/ian8667/powershell/commit/a34a32799535e61e73e38077e18550b71cafb6ba) from last year. I originally tested it in PowerShell 7 as that's my default shell.

I did test it again using PowerShell 5, and it behaves the same. The issue is that the script assumes that moving the file doesn't change where on the disk it's stored, so the fault is easily replicated by deleting the file on another volume.

You can create and mount a virtual volume in disk management if you want to test it yourself:
![image](https://github.com/ian8667/powershell/assets/93410187/1af02200-9e7b-478f-ac48-4ea7922323bc)

Here's a screenshot of me running this test in Windows PowerShell (PS5). I ran recuva on D:\ and had it save the file to my downloads folder before I ran cat at the end:

![image](https://github.com/ian8667/powershell/assets/93410187/0ae0ec13-b0cb-4708-93e0-86acf785d270)

**Comment by ian8667** [comment]
Hi,
Has this script ever worked successfully for you?

**Comment by Daigle-ISM** [comment]
The issue does not occur with the commit previous to the one I referenced:
https://github.com/ian8667/powershell/blob/0ade56766965906df716eaf314e78682b8497627/Secure-Delete.ps1

Using that version the files deleted, even on separate volumes, are not recovered by Recuva.

**Comment by ian8667** [comment]
Hi,
Just to let you know that I've taken your comments on board and I'm rewriting the program so as to overwrite the target file in it's original directory.
I'll let you know when I feel it's complete.
Thanks

**Comment by ian8667** [comment]
Hi,
I've uploaded file 'Secure-Delete.ps1' version '2023-07-13T16:56:53'. Can you try this version please? In particular, see whether you can use Recuva to recover the file.
Can you let me know how you get on please?
Thanks.

**Comment by Daigle-ISM** [comment]
Thanks for that. I've been trying this out. It seems to resolve the original issue, but there's a new problem I've encountered.

I'm encountering what appears to be a race condition with the stream. All of the stream methods being used are synchronous, but the data in the stream isn't actually getting written to the file unless I insert a breakpoint or a long enough sleep to slow the loop down. Clearly I don't understand streams as well as I had thought.

This means the final Remove-Item is deleting the file while it still has its original contents, which I can recover every time (unless there's a sleep or breakpoint after the flush() methods).

I replicated this issue with this very basic script (in PS5 and PS7):

```powershell
[CmdletBinding()]
param (
    [Parameter(Mandatory=$true, Position=0)]
    [string]
    $Path
)
$File = Get-Item $Path
$RNG = [System.Security.Cryptography.RandomNumberGenerator]::Create()
$Bytes = [Byte[]]::New($File.Length)
$RNG.GetBytes($Bytes)
$Stream = $File.OpenWrite()

$Stream.Write($Bytes, 0, $Bytes.Length)
$Stream.Flush()
$Stream.Close()
Remove-Item $File
```

And in C# as well (yes, this issue even occurred when I initialized the stream with a using block):

```csharp
Console.WriteLine("Enter the path to the file you want to destroy: ");
string path = Console.ReadLine();
FileInfo file = new FileInfo(path);
var rng = System.Security.Cryptography.RandomNumberGenerator.Create();
byte[] bytes = new byte[file.Length];
rng.GetBytes(bytes);
var stream = file.OpenWrite();

stream.Write(bytes, 0, bytes.Length);
stream.Flush();
stream.Close();
file.Delete();
```

I took a peek at the source code behind [Eraser](https://eraser.heidi.ie/) and it seems to use a similar strategy with the filestreams, though with a LOT more logic and cases handled. After looking at what they've done and all of the considerations given, I have to think that using their tool is probably a better option.

**Comment by ian8667** [comment]
Hi,
When you say you've looked at the source code behind [Eraser], are you able to give me a quick look at this please?
Thanks.

**Comment by Daigle-ISM** [comment]
Their source control system is a bit tricky, I got it from here:
<https://sourceforge.net/p/eraser/code/HEAD/tree/trunk/eraser/>

Definitely worth a look. The section that performs a similar function to your code is in the `Erase()` method at `eraser\Eraser.DefaultPlugins\ErasureMethods\ErasureMethodBase.cs:171`

Looking a bit closer at how they're doing it (namely in ntfs.cs where the stream is opened), we're missing the `FileOptions.WriteThrough` option. The below example works as expected:

```powershell
[CmdletBinding()]
param (
    [Parameter(Mandatory=$true, Position=0)]
    [string]
    $Path
)
$File = Get-Item $Path
$RNG = [System.Security.Cryptography.RandomNumberGenerator]::Create()
$Bytes = [Byte[]]::New($File.Length)
$RNG.GetBytes($Bytes)
[System.IO.FileStreamOptions]$FSOptions = @{
    Mode = [System.IO.FileMode]::Open
    Access = [System.IO.FileAccess]::Write
    Share = [System.IO.FileShare]::None
    Options = [System.IO.FileOptions]::WriteThrough
}
$Stream = $File.Open($FSOptions)

$Stream.Write($Bytes, 0, $Bytes.Length)
$Stream.Flush()
$Stream.Close()
Remove-Item $File
```

There may be (are probably) a lot of other nuances of which I'm not aware, Eraser's code may provide more insight

**Comment by ian8667** [comment]
Hi,
Thanks for the information. Taking your comments onboard, I've modified the code to ensure all writes to the file stream have completed before the file is deleted with the cmdlet ‘Remove-Item’. Because I'm using the method <System.IO.FileStream>.WriteAsync to write to the stream, I use the method <System.Threading.Tasks.Task>.Wait() to wait for the Task to complete execution.

To ensure that all buffered data in intermediate file buffers is written to disk, I used the method <System.IO.FileStream>.Flush($true).

I've uploaded version '2023-08-14T16:04:10' of the file 'Secure-Delete.ps1'. Grateful if you will have a go with this version to see if it's any better for you when you get time.

Thanks.
Regards
